rust/validation/src/validation/dict.rs

```diff
@@ -2,116 +2,128 @@
         if let Some(ref_) = self.base.schema_ref.as_ref() {
             // Ignoring not being able to resolve the schema.
             // Ignoring a wrong schema type at the ref. Since Validation is infallible.
             // TODO: What to do?
             if let Ok(AnySchema::Dict(ref_schema)) = resolve_ref(ref_, ctx.store) {
+                // Handle relaxed validation here, since the places we use it is also where we skip resolving the $ref before validation.
+                let previous_relaxed_validation = ctx.state.relaxed_validation;
+                if self.relaxed_validation.unwrap_or_default() {
+                    ctx.state.relaxed_validation = true
+                }
                 ref_schema.validate(value, ctx);
+                ctx.state.relaxed_validation = previous_relaxed_validation;
             }
         }
     }
-
-    fn default_value(&self) -> Option<Map<String, Value>> {
-        if let Some(value) = &self.base.default {
-            let value = value.to_owned();
-            let map = Map::from_iter(value);
-            Some(map)
-        } else {
-            None
-        }
-    }
 }
 
-fn validate_allowed_keys(schema: &Dict, input: &Map<String, Value>, ctx: &mut Context) {
-    if !schema.allow_other_keys.unwrap_or_default()
-        && let Some(keys) = &schema.keys {
-            input
-                .keys()
-                // keys starting with "_" are passed over to allow for custom usage
-                .filter(|key| !key.starts_with('_'))
-                .filter(|key| !keys.contains_key(key.as_str()))
-                .for_each(|key| {
-                    ctx.path.push(key.to_owned());
-                    ctx.add_violation(Violation::UnexpectedKey());
-                    ctx.path.pop();
-                });
-        }
+fn get_dynamic_keys_schemas<'a>(
+    schema: &'a Dict,
+    input: &'a Map<String, Value>,
+) -> OrderMap<String, &'a AnySchema> {
+    schema
+        .get_dynamic_keys(input)
+        .into_iter()
+        .flat_map(|dynamic_keys| {
+            dynamic_keys
+                .into_iter()
+                .map(|(key, dynamic_key_info)| (key, dynamic_key_info.schema))
+        })
+        .collect()
 }
 
 fn validate_keys(schema: &Dict, input: &Map<String, Value>, ctx: &mut Context) {
     if let Some(keys) = &schema.keys {
-        for (key, key_schema) in keys {
-            match input.get(key) {
-                Some(Value::Null) | None => {
-                    // nullish values don't need to be validated beyond a requiredness check
+        let dynamic_keys_schemas = get_dynamic_keys_schemas(schema, input);
+        input.iter().for_each(|(input_key, input_value)| {
+            ctx.state.path.push(input_key.to_owned());
+            if let Some(key_schema) = keys.get(input_key) {
+                check_deprecation(input_key, key_schema, input, ctx);
+                key_schema.validate(input_value, ctx);
```

Comment:
so here if check_deprecation is adding an error, we still validate the key? Often we don't have type information anymore right?

Reply:
we *always* have the type, but you are correct, if it is removed or a conflict we should stop there. I will work on this.